File: src/tracing.js
mergeInto(LibraryManager.library, {
  emscripten_trace_configure: function(collector_url, app_name) {
    console.log("Custom tracing up", collector_url, app_name);
  },
  emscripten_trace_record_frame_start: function() {
    console.log("TRACE: frame start");
  },
  emscripten_trace_record_frame_end: function() {
    console.log("TRACE: frame end");
  },
  emscripten_trace_report_memory_layout__deps: ['emscripten_stack_get_base', 'emscripten_stack_get_current', 'emscripten_stack_get_end', 'sbrk'],
  emscripten_trace_report_memory_layout: function() {
    console.log("TRACE: memory layout\n", {
      'static_base':  {{{ GLOBAL_BASE }}},
      'stack_base':   _emscripten_stack_get_base(),
      'stack_top':    _emscripten_stack_get_current(),
      'stack_max':    _emscripten_stack_get_end(),
      'dynamic_top':  _sbrk(0),
      'total_memory': HEAP8.length
    });
  },
  emscripten_trace_report_off_heap_data: function() {},
  emscripten_trace_close: function() {},
  emscripten_trace_record_allocation: function (address, size) {
    console.log("TRACE: alloc", address, size);
  },
  emscripten_trace_record_reallocation: function (old_address, new_address, size) {
    console.log("TRACE: realloc", address, size);
  },
  emscripten_trace_record_free: function (address) {
    console.log("TRACE: free", address);
  },
  emscripten_trace_log_message: function (channel, message) {
    console.log("TRACE: message", UTF8ToString(channel), UTF8ToString(message));
  }
});

File: src/tracing.cpp
#include "tracing.hpp"

#ifdef __EMSCRIPTEN__

#include <emscripten/trace.h>

void configureTracing(void) {
  emscripten_trace_configure("http://127.0.0.1:5000/", "openjphjs");
}

void reportFrameStart() {
  emscripten_trace_record_frame_start();
}

void reportFrameEnd() {
  emscripten_trace_record_frame_end();
}

void reportMemoryUsage() {
  emscripten_trace_report_memory_layout();
  emscripten_trace_report_off_heap_data();
}

void reportMessage(const char *channel, const char *message) {
  emscripten_trace_log_message(channel, message);
}

void closeTracing(void) {
  emscripten_trace_close();
}

#else

void configureTracing(void) {}
void reportFrameStart(void) {};
void reportFrameEnd(void) {};
void reportMemoryUsage() {}
void reportMessage(const char *channel, const char *message) {}
void closeTracing(void) {}

#endif
